File: CourseProjects/CourseProjects/VisualStudioProjects/Agile/MockingWithMoq/AdvancedMocking/ShopTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;

namespace AdvancedMocking {
    [TestClass]
    public class ShopTest {
        [TestInitialize]
        public void start() {
            pricingMock = new Mock<IPricingEngine>();
            paymentMock = new Mock<IPaymentEngine>();
            stockCheckMock = new Mock<IStockCheckEngine>();
        }
        [TestMethod]
        public void MakePurchaseWorksForValidQuantity() {
            stockCheckMock.Setup(m => m.Check("AB12")).Returns(20).Verifiable();
            pricingMock.Setup(m => m.Price("AB12", 19)).Returns(27.30).Verifiable();
            paymentMock.Setup(m => m.Authorize("010203XYZ", 27.30)).Returns(true).Verifiable();

            Shop s = new Shop(pricingMock.Object, stockCheckMock.Object, paymentMock.Object);
            Assert.IsTrue(s.MakePurchase("AB12", 19, "010203XYZ"));

            stockCheckMock.Verify();
            pricingMock.Verify();
            paymentMock.Verify();
        }
        [TestMethod]
        public void MakePurchaseWorksForInvalidQuantity() {
            stockCheckMock.Setup(m => m.Check("AB12")).Returns(20).Verifiable();

            Shop s = new Shop(pricingMock.Object, stockCheckMock.Object, paymentMock.Object);
            Assert.IsFalse(s.MakePurchase("AB12", 21, "010203XYZ"));

            stockCheckMock.Verify();
        }
        [TestMethod]
        public void OrderPossibleWhenItemsInStock() {
            var items = new SortedDictionary<string, int>();
            items.Add("AB12", 12);
            items.Add("CD34", 34);
            items.Add("EF56", 56);

            stockCheckMock.Setup(m => m.Check(It.IsAny<string>())).Returns(100);

            Shop shop = new Shop(pricingMock.Object, stockCheckMock.Object, paymentMock.Object);

            Assert.AreEqual(0, shop.CheckForUnderstockedItems(items).Count);
            stockCheckMock.Verify(m => m.Check(It.IsAny<string>()), Times.Exactly(3));
        }
        [TestMethod]
        public void OrderNotPossibleWhenSomeItemsOutOfStock() {
            var items = new SortedDictionary<string, int>();
            items.Add("AB12", 12);
            items.Add("CD34", 34);
            items.Add("EF56", 56);

            stockCheckMock.Setup(m => m.Check("AB12")).Returns(14);
            stockCheckMock.Setup(m => m.Check("CD34")).Returns(31);
            stockCheckMock.Setup(m => m.Check("EF56")).Returns(50);

            Shop shop = new Shop(pricingMock.Object, stockCheckMock.Object, paymentMock.Object);
            var results = shop.CheckForUnderstockedItems(items);

            stockCheckMock.Verify(m => m.Check(It.IsAny<string>()), Times.Exactly(3));
            Assert.AreEqual(2, results.Count);
            Assert.AreEqual(3, results["CD34"]);
            Assert.AreEqual(6, results["EF56"]);
        }
        [TestMethod]
        public void OrderNotPossibleWhenAllItemsOutOfStock() {
            var items = new SortedDictionary<string, int>();
            items.Add("AB12", 12);
            items.Add("CD34", 34);
            items.Add("EF56", 56);

            int count = 10;

            stockCheckMock.Setup(m => m.Check(It.IsAny<string>())).Returns((string s) => count++);

            Shop shop = new Shop(pricingMock.Object, stockCheckMock.Object, paymentMock.Object);
            var results = shop.CheckForUnderstockedItems(items);

            Assert.AreEqual(3, results.Count);
            Assert.AreEqual(2, results["AB12"]);
            Assert.AreEqual(23, results["CD34"]);
            Assert.AreEqual(44, results["EF56"]);
        }
        [TestMethod]
	    public void MakeMultiplePurchasesWorksForValidQuantities() {
            var items = new SortedDictionary<string, int>();
            items.Add("AB12", 12);
            items.Add("CD34", 34);
            items.Add("EF56", 56);

            stockCheckMock.Setup(m => m.Check(It.IsAny<string>())).Returns(100);

            pricingMock.Setup(m => m.Price("AB12", 12)).Returns(1.23);
            pricingMock.Setup(m => m.Price("CD34", 34)).Returns(3.45);
            pricingMock.Setup(m => m.Price("EF56", 56)).Returns(6.78);

            paymentMock.Setup(m => m.Authorize("010203XYZ", 11.46)).Returns(true).Verifiable();

            Shop shop = new Shop(pricingMock.Object, stockCheckMock.Object, paymentMock.Object);
		    Assert.AreEqual(PurchaseResults.SUCCESS,shop.MakePurchases(items, "010203XYZ"));

            stockCheckMock.Verify(m => m.Check(It.IsAny<string>()), Times.Exactly(3));
            pricingMock.Verify(m => m.Price(It.IsAny<string>(), It.IsAny<int>()),Times.Exactly(3));
            paymentMock.Verify();
	    }
        [TestMethod]
        public void MakeMultiplePurchasesFailsForInvalidQuantities() {
            var items = new SortedDictionary<string, int>();
            items.Add("AB12", 12);
            items.Add("CD34", 34);
            items.Add("EF56", 56);

            stockCheckMock.Setup(m => m.Check(It.IsAny<string>())).Returns(1);

            Shop shop = new Shop(pricingMock.Object, stockCheckMock.Object, paymentMock.Object);
            Assert.AreEqual(PurchaseResults.ITEMS_OUT_OF_STOCK, shop.MakePurchases(items, "010203XYZ"));

            stockCheckMock.Verify(m => m.Check(It.IsAny<string>()), Times.Exactly(3));
            pricingMock.Verify(m => m.Price(It.IsAny<string>(), It.IsAny<int>()), Times.Never());
            paymentMock.Verify(m => m.Authorize(It.IsAny<string>(), It.IsAny<double>()), Times.Never());
        }
        [TestMethod]
        public void MakeMultiplePurchasesFailsWhenCardInvalid() {
            var items = new SortedDictionary<string, int>();
            items.Add("AB12", 12);
            items.Add("CD34", 34);
            items.Add("EF56", 56);

            stockCheckMock.Setup(m => m.Check(It.IsAny<string>())).Returns(100);
            pricingMock.Setup(m => m.Price(It.IsAny<string>(), It.IsAny<int>())).Returns(2.0);
            paymentMock.Setup(m => m.Authorize("010203XYZ", 6.0)).Throws(new InvalidCardException());

            Shop shop = new Shop(pricingMock.Object, stockCheckMock.Object, paymentMock.Object);
            Assert.AreEqual(PurchaseResults.CARD_INVALID, shop.MakePurchases(items, "010203XYZ"));
        }
        private Mock<IPricingEngine> pricingMock;
        private Mock<IPaymentEngine> paymentMock;
        private Mock<IStockCheckEngine> stockCheckMock;
    }
}

File: CourseProjects/CourseProjects/JavaEclipseProjects/UnitTestingExamples/src/demos/tdd/mockito/advanced/ShopTest.java
package demos.tdd.mockito.advanced;

import java.util.SortedMap;
import java.util.TreeMap;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import static org.mockito.Mockito.*;

public class ShopTest {
	@Before
	public void start() {
		pricingEngine = mock(PricingEngine.class);
		paymentEngine = mock(PaymentEngine.class);
		stockCheckEngine = mock(StockCheckEngine.class);
	}
	@Test
	public void makePurchaseWorksForValidQuantity() {
		when(stockCheckEngine.check("AB12")).thenReturn(20);
		when(pricingEngine.price("AB12",19)).thenReturn(27.30);
		when(paymentEngine.authorize("010203XYZ",27.30)).thenReturn(true);
		
		Shop s = new Shop(pricingEngine,stockCheckEngine,paymentEngine);
		assertTrue(s.makePurchase("AB12", 19, "010203XYZ"));
		
		verify(stockCheckEngine).check("AB12");
		verify(pricingEngine).price("AB12",19);
		verify(paymentEngine).authorize("010203XYZ",27.30);
	}
	@Test
	public void makePurchaseWorksForInvalidQuantity() {
		when(stockCheckEngine.check("AB12")).thenReturn(20);

		Shop s = new Shop(pricingEngine,stockCheckEngine,paymentEngine);
		assertFalse(s.makePurchase("AB12", 21, "010203XYZ"));

		verify(stockCheckEngine).check("AB12");
	}
	@Test
	public void orderPossibleWhenItemsInStock() {
		SortedMap<String,Integer> items = new TreeMap<String,Integer>();
		items.put("AB12", 12);
		items.put("CD34", 34);
		items.put("EF56", 56);
		
		when(stockCheckEngine.check("AB12")).thenReturn(14);
		when(stockCheckEngine.check("CD34")).thenReturn(36);
		when(stockCheckEngine.check("EF56")).thenReturn(58);
		
		Shop shop = new Shop(pricingEngine,stockCheckEngine,paymentEngine);
		assertEquals(0,shop.checkIfOrderPossible(items).length);
		
		verify(stockCheckEngine,times(3)).check(anyString());
	}
	@Test
	public void orderNotPossibleWhenItemsOutOfStock() {
		SortedMap<String,Integer> items = new TreeMap<String,Integer>();
		items.put("AB12", 12);
		items.put("CD34", 34);
		items.put("EF56", 56);
		
		when(stockCheckEngine.check("AB12")).thenReturn(14);
		when(stockCheckEngine.check("CD34")).thenReturn(32);
		when(stockCheckEngine.check("EF56")).thenReturn(54);
		
		Shop shop = new Shop(pricingEngine,stockCheckEngine,paymentEngine);
		String [] results = shop.checkIfOrderPossible(items);
		
		verify(stockCheckEngine,times(3)).check(anyString());
		assertEquals(2,results.length);
		assertEquals("CD34",results[0]);
		assertEquals("EF56",results[1]);
	}
	@Test
	public void makeMultiplePurchasesWorksForValidQuantities() {
		SortedMap<String,Integer> items = new TreeMap<String,Integer>();
		items.put("AB12", 12);
		items.put("CD34", 34);
		items.put("EF56", 56);
		
		when(stockCheckEngine.check(anyString())).thenReturn(100);
		
		when(pricingEngine.price("AB12",12)).thenReturn(1.23);
		when(pricingEngine.price("CD34",34)).thenReturn(3.45);
		when(pricingEngine.price("EF56",56)).thenReturn(6.78);
		
		when(paymentEngine.authorize("010203XYZ",11.46)).thenReturn(true);
		
		Shop shop = new Shop(pricingEngine,stockCheckEngine,paymentEngine);
		assertTrue(shop.makePurchases(items, "010203XYZ"));
		
		verify(stockCheckEngine,times(3)).check(anyString());
		verify(pricingEngine,times(3)).price(anyString(),anyInt());
		verify(paymentEngine).authorize("010203XYZ",11.46);
	}
	private PricingEngine pricingEngine;
	private PaymentEngine paymentEngine;
	private StockCheckEngine stockCheckEngine;
}
